### PR TITLE
forbid invalid edge weights and negative default weight in K_SHORTEST

### PR DESCRIPTION
**Pull request by hkernbach**
3.8 backport of: https://github.com/arangodb/arangodb/pull/14218
Please see description there.

### CODE REVIEW

File: arangod/Graph/ShortestPathOptions.cpp [hunk 1]
```diff
@@ -162,8 +160,13 @@ void ShortestPathOptions::addReverseLookupInfo(aql::ExecutionPlan* plan,
 
 double ShortestPathOptions::weightEdge(VPackSlice edge) const {
   TRI_ASSERT(useWeight());
-  return arangodb::basics::VelocyPackHelper::getNumericValue<double>(
-      edge, weightAttribute.c_str(), defaultWeight);
+  const auto weight = arangodb::basics::VelocyPackHelper::getNumericValue<double>(
+      edge, _weightAttribute.c_str(), _defaultWeight);
```
Comment (maierlars) [inline arangod/Graph/ShortestPathOptions.cpp:164]: should work without `c_str`.

File: arangod/Graph/ShortestPathOptions.cpp [hunk 2]
```diff
@@ -212,18 +215,41 @@ auto ShortestPathOptions::estimateDepth() const noexcept -> uint64_t {
...
+  }
+  _defaultWeight = weight;
+}
+
+auto ShortestPathOptions::setWeightAttribute(std::string attribute) -> void {
+  _weightAttribute = attribute;
+}
+
+auto ShortestPathOptions::getDefaultWeight() const -> double {
+  TRI_ASSERT(_defaultWeight > .0);
```
Comment (maierlars) [inline arangod/Graph/ShortestPathOptions.cpp:231]: `>=` ?

File: arangod/Graph/ShortestPathOptions.cpp [hunk 3]
```diff
@@ -212,18 +215,41 @@ auto ShortestPathOptions::estimateDepth() const noexcept -> uint64_t {
...
       end{other.end},
-      weightAttribute{other.weightAttribute},
-      defaultWeight{other.defaultWeight},
       bidirectional{other.bidirectional},
       multiThreaded{other.multiThreaded},
-      _reverseLookupInfos{other._reverseLookupInfos} {}
+      _reverseLookupInfos{other._reverseLookupInfos},
+      _weightAttribute{other._weightAttribute},
+      _defaultWeight{other._defaultWeight} {
+  TRI_ASSERT(other._defaultWeight >= 0.);
```
Comment (maierlars) [inline arangod/Graph/ShortestPathOptions.cpp:251]: `>`?

**Review by maierlars — COMMENTED** [review]
I think we should go with `> 0.` for shortest path and k shortest path.

File: arangod/Graph/ShortestPathOptions.cpp [hunk 4]
```diff
@@ -162,8 +160,13 @@ void ShortestPathOptions::addReverseLookupInfo(aql::ExecutionPlan* plan,
 
 double ShortestPathOptions::weightEdge(VPackSlice edge) const {
   TRI_ASSERT(useWeight());
-  return arangodb::basics::VelocyPackHelper::getNumericValue<double>(
-      edge, weightAttribute.c_str(), defaultWeight);
+  const auto weight = arangodb::basics::VelocyPackHelper::getNumericValue<double>(
+      edge, _weightAttribute.c_str(), _defaultWeight);
+  if (weight < 0.) {
```
Comment (maierlars) [inline arangod/Graph/ShortestPathOptions.cpp:165]: `<= 0.`

File: arangod/Graph/ShortestPathOptions.cpp [hunk 5]
```diff
@@ -212,18 +217,42 @@ auto ShortestPathOptions::estimateDepth() const noexcept -> uint64_t {
...
+auto ShortestPathOptions::setDefaultWeight(double weight) -> void {
+  if (weight <= 0.) {
+    THROW_ARANGO_EXCEPTION_MESSAGE(
+        TRI_ERROR_GRAPH_INVALID_EDGE_WEIGHT,
+        "negative or zero value default weight not allowed");
+  }
+  _defaultWeight = weight;
+}
+
+auto ShortestPathOptions::setWeightAttribute(std::string attribute) -> void {
```
Comment (gnusi) [inline arangod/Graph/ShortestPathOptions.cpp:229]: I suggest we avoid unnecessary string copy

File: arangod/Graph/ShortestPathOptions.cpp [hunk 6]
```diff
@@ -212,18 +217,42 @@ auto ShortestPathOptions::estimateDepth() const noexcept -> uint64_t {
...
+  _weightAttribute = attribute;
+}
+
+auto ShortestPathOptions::getDefaultWeight() const -> double {
+  TRI_ASSERT(_defaultWeight > .0);
+  return _defaultWeight;
+}
+
+auto ShortestPathOptions::getWeightAttribute() const -> std::string {
```
Comment (gnusi) [inline arangod/Graph/ShortestPathOptions.cpp:238]: I suggest we avoid unnecessary string copy

**Review by gnusi — CHANGES_REQUESTED** [review]

File: arangod/Graph/ShortestPathOptions.cpp (hunk 5 above)
Comment (hkernbach) [reply arangod/Graph/ShortestPathOptions.cpp:226]: Done in: eebdeb77e47ac1e4ad2e86e587288a3c0cc4da28
(also included in devel port)

File: arangod/Graph/ShortestPathOptions.cpp (hunk 6 above)
Comment (hkernbach) [reply arangod/Graph/ShortestPathOptions.cpp]: Done in: eebdeb77e47ac1e4ad2e86e587288a3c0cc4da28
(also included in devel port)

**Review by gnusi — APPROVED** [review]

**Pull request merged**